sync/src/main/java/tech/pegasys/artemis/sync/PendingBlocks.java

```diff
@@ -0,0 +1,173 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.sync;
+
+import com.google.common.eventbus.EventBus;
+import com.google.common.eventbus.Subscribe;
+import com.google.common.primitives.UnsignedLong;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.service.serviceutils.Service;
+import tech.pegasys.artemis.storage.events.SlotEvent;
+import tech.pegasys.artemis.util.config.Constants;
+
+class PendingBlocks extends Service {
+  private static final UnsignedLong DEFAULT_FUTURE_BLOCK_TOLERANCE = UnsignedLong.valueOf(2);
+  private static final UnsignedLong DEFAULT_HISTORICAL_BLOCK_TOLERANCE =
+      UnsignedLong.valueOf(Constants.SLOTS_PER_EPOCH * 10);
+  private static final UnsignedLong GENESIS_SLOT = UnsignedLong.valueOf(Constants.GENESIS_SLOT);
+
+  private final EventBus eventBus;
+  private final Map<Bytes32, BeaconBlock> pendingBlocks = new ConcurrentHashMap<>();
+  private final Map<Bytes32, Set<Bytes32>> pendingBlocksByParentRoot = new ConcurrentHashMap<>();
+  // Define the range of slots we care about
+  private final UnsignedLong futureBlockTolerance;
+  private final UnsignedLong historicalBlockTolerance;
+
+  private volatile UnsignedLong currentSlot = UnsignedLong.ZERO;
+
+  PendingBlocks(
+      final EventBus eventBus,
+      final UnsignedLong historicalBlockTolerance,
+      final UnsignedLong futureBlockTolerance) {
+    this.eventBus = eventBus;
+    this.historicalBlockTolerance = historicalBlockTolerance;
+    this.futureBlockTolerance = futureBlockTolerance;
+  }
+
+  public static PendingBlocks create(final EventBus eventBus) {
+    return new PendingBlocks(
+        eventBus, DEFAULT_HISTORICAL_BLOCK_TOLERANCE, DEFAULT_FUTURE_BLOCK_TOLERANCE);
+  }
+
+  @Override
+  protected CompletableFuture<?> doStart() {
+    eventBus.register(this);
+    return CompletableFuture.completedFuture(null);
+  }
+
+  public void add(BeaconBlock beaconBlock) {
+    if (shouldIgnoreBlock(beaconBlock)) {
+      // Ignore blocks outside of the range we care about
+      return;
+    }
+
+    final Bytes32 blockRoot = beaconBlock.signing_root("signature");
+    final Bytes32 parentRoot = beaconBlock.getParent_root();
+
+    // Index block by parent
+    pendingBlocksByParentRoot
+        .computeIfAbsent(parentRoot, (key) -> createRootSet(blockRoot))
+        .add(blockRoot);
```

Comment:
`createRootSet` is already adding `blockRoot` so we wind up adding it twice (which works out because it's a set but is unnecessary)

Reply:
Adding twice just to make sure we don't accidentally prune the empty set before the first root is added (empty sets are pruned in the `remove` call).  I'll add a comment.

Reply:
Ahhh, good thinking - I totally missed that.